### PR TITLE
Avoid multiple enumerations of an `IEnumerable<Expression>` instance in `Compiler.cs`

### LINKED ISSUE #13474: LINQ in the compiler
Not really a bug but:

https://github.com/PowerShell/PowerShell/blob/904e5511789f0a8375c23c9a1ee59918be0a82a1/src/System.Management.Automation/engine/parser/Compiler.cs#L952-L955

And the method in the select:

https://github.com/PowerShell/PowerShell/blob/904e5511789f0a8375c23c9a1ee59918be0a82a1/src/System.Management.Automation/engine/parser/Compiler.cs#L941-L950

That method is used to compile method arguments, the result is then enumerated twice (once in the first highlighted line, one is the last):

https://github.com/PowerShell/PowerShell/blob/904e5511789f0a8375c23c9a1ee59918be0a82a1/src/System.Management.Automation/engine/parser/Compiler.cs#L6338-L6344

This causes all method arguments to be compiled twice due to multiple enumeration.  Also the method supplied to `Select` is not done in such a way that the delegate can be cached per instance by roslyn.

### PR DESCRIPTION
**Pull request by daxian-dbw**
# PR Summary

Fix #13474
Avoid multiple enumerations of an `IEnumerable<Expression>` instance at a few places in `Compiler.cs`.
More than one enumeration cause the operations on individual elements of the source collection to be done repeatedly.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

**Review by rjmholt — APPROVED** [review]
Seems like a decent fix for now. It would be nice if we could remove `using System.Linq;` from the top of Compiler.cs and clean up all the LINQ in there

File: src/System.Management.Automation/engine/parser/Compiler.cs [hunk 1]
```diff
@@ -951,7 +951,7 @@ internal Expression CompileExpressionOperand(ExpressionAst exprAst)
 
         private IEnumerable<Expression> CompileInvocationArguments(IEnumerable<ExpressionAst> arguments)
         {
-            return arguments == null ? Array.Empty<Expression>() : arguments.Select(CompileExpressionOperand);
+            return arguments == null ? Array.Empty<Expression>() : arguments.Select(CompileExpressionOperand).ToArray();
```
Comment (SeeminglyScience) [inline src/System.Management.Automation/engine/parser/Compiler.cs:954]: Looks like everywhere this method is called passes a `ReadOnlyCollection<ExpressionAst>`.  Could type `arguments` as `IReadOnlyList` and do this:

```csharp
if (arguments is null)
{
    return Array.Empty<Expression>();
}

var result = new Expression[arguments.Count];
for (int i = result.Length - 1; i >= 0; i--)
{
    result[i] = CompileExpressionOperand(arguments[i]);
}

return result;
```

If you would rather not change the method parameter type then LGTM!

**Review by SeeminglyScience — APPROVED** [review]

**Comment by vexx32** [comment]
@rjmholt is that an invitation? 😏

File: src/System.Management.Automation/engine/parser/Compiler.cs (hunk 1 above)
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/parser/Compiler.cs]: Code updated.

**Comment by rjmholt** [comment]
@poshchan please remind me in 20 hours

**Review by iSazonov — APPROVED** [review]
Oh, it looks as a trick because these methods (InvokeMember/InvokeBaseCtorMethod/InvokeDynamicMember) still use IEnumerable<ExpressionAst> type.

**Pull request merged**